docs/dev/framework/routing/content-routing.md

```diff
@@ -0,0 +1,301 @@
+shown within a certain [page controller][PageControllers] you can generate the parameters that your page controller's
+route uses for the given content.
+
+
+### Custom Example
+
+The follwing shows a custom example using a page controller. Let's say you have a custom DCA `tl_foobar` and an 
+accompanying `FoobarModel`:
+
+```php
+// contao/dca/tl_foobar.php
+$GLOBALS['TL_DCA']['tl_foobar'] = […];
+```
+
+```php
+namespace App\Model;
+
+use Contao\Model;
+
+class FoobarModel extends Model
+{
+    protected static $strTable = 'tl_foobar';
+}
+```
+
+```php
+use App\Model\FoobarModel;
+
+$GLOBALS['BE_MOD']['content']['Foobar'] = ['tables' => ['tl_foobar']];
+$GLOBALS['TL_MODELS']['tl_foobar'] = FoobarModel::class;
+```
+
+You want to render the details of these records via your own page controller and front end module:
+
+```php
+// src/Controller/Page/FoobarReaderController.php
+namespace App\Controller\Page;
+
+use App\Model\FoobarModel;
+use Contao\CoreBundle\DependencyInjection\Attribute\AsPage;
+use Contao\CoreBundle\Exception\PageNotFoundException;
+use Contao\FrontendIndex;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+#[AsPage(path: '{foobarId}', requirements: ['foobarId' => '\d+'])]
+class FoobarReaderController
+{
+    public function __invoke(Request $request, PageModel $pageModel, int $foobarId): Response
+    {
+        if (!$record = FoobarModel::findById($foobarId)) {
+            throw new PageNotFoundException();
+        }
+
+        $request->attributes->set('_content', $record);
+
+        return (new FrontendIndex())->renderPage($pageModel);
+    }
+}
+```
+
+Now our resolver to generate URLs to these records could look like this:
+
+```php
+// src/Routing/FoobarResolver.php
+namespace App\Routing;
+
+use App\Model\FoobarModel;
+use Contao\CoreBundle\Routing\Content\ContentUrlResolverInterface;
+use Contao\CoreBundle\Routing\Content\ContentUrlResult;
+use Contao\CoreBundle\Routing\PageFinder;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+class FoobarResolver implements ContentUrlResolverInterface
+{
+    public function __construct(
+        private readonly RequestStack $requestStack,
+        private readonly PageFinder $pageFinder,
+    ) {
+    }
+
+    public function resolve(object $content): ContentUrlResult|null
+    {
+        if (!$content instanceof FoobarModel) {
+            return null;
+        }
+
+        if (!$request = $this->requestStack->getCurrentRequest()) {
+            return null;
+        }
+
+        /**
+         * This is a simplification for this example. We simply look for the first "foobar_reader" page type in this
+         * website and assume that this is the correct page for which we want to generate the detail URL of our record.
+         */
+        $foobarPage = $this->pageFinder->findFirstPageOfTypeForRequest($request, 'foobar_reader');
```

Comment:
I would simplify this e.g. by using a property in `$content` as the ID of the target page, removing the complexity of request and pageFinder.

Reply:
That would make the example more complex actually as you would then have to define the reader page somewhere. Either per record (which would be silly) or in a parent record (which this example does not use, i.e. it does not use mode 4).

This is actually how I would implement it in the new `contao/blog-bundle` - you don't _have to_ define a redirect page in the blog archive anymore. If you don't, the system will just look for the first reader page in the page tree. This simplifies the usage.